ktx2Decoder/src/Transcoders/liteTranscoder.ts

```diff
@@ -0,0 +1,70 @@
+import { Nullable } from '../types';
+import { Transcoder, sourceTextureFormat, transcodeTarget } from '../transcoder';
+import { WASMMemoryManager } from '../wasmMemoryManager';
+import { KTX2FileReader, IKTX2_ImageDesc } from '../ktx2FileReader';
+
+/**
+ * @hidden
+ */
+export class LiteTranscoder extends Transcoder {
+
+    private _modulePath: string;
+    private _modulePromise: Promise<{ module: any }>;
+    private _memoryManager: WASMMemoryManager;
+
+    protected _loadModule(): Promise<{ module: any }> {
+        if (this._modulePromise) {
+            return this._modulePromise;
+        }
+
+        this._modulePromise = new Promise((resolve, reject) => {
+            WASMMemoryManager.LoadWASM(this._modulePath)
+                .then((wasmBinary) => {
+                    WebAssembly.instantiate(wasmBinary as ArrayBuffer, { env: { memory: this._memoryManager.wasmMemory } }).then((moduleWrapper) => {
+                        resolve({ module: moduleWrapper.instance.exports });
+                    });
+                }, (reason) => {
+                    reject(reason);
+                })
+                .catch((reason) => {
+                    reject(reason);
+                });
+        });
+
+        return this._modulePromise;
+    }
+
+    protected get memoryManager(): WASMMemoryManager {
+        return this._memoryManager;
+    }
+
+    protected setModulePath(modulePath: string): void {
+        this._modulePath = modulePath;
+    }
+
+    public needMemoryManager(): boolean {
+        return true;
+    }
+
+    public setMemoryManager(memoryMgr: WASMMemoryManager): void {
+        this._memoryManager = memoryMgr;
+    }
+
+    public transcode(src: sourceTextureFormat, dst: transcodeTarget, level: number, width: number, height: number, uncompressedByteLength: number, ktx2Reader: KTX2FileReader, imageDesc: Nullable<IKTX2_ImageDesc>, encodedData: Uint8Array): Promise<Nullable<Uint8Array>> {
+        return this._loadModule().then((moduleWrapper: any) => {
+            const transcoder: any = moduleWrapper.module;
+
+            const nBlocks = ((width + 3) >> 2) * ((height + 3) >> 2);
+
+            const texMemoryPages = ((nBlocks * 16 + 65535) >> 16) + 1;
+
+            const textureView = this.memoryManager.getMemoryView(texMemoryPages, 65536, nBlocks * 16);
+
+            textureView.set(encodedData);
+
+            return transcoder.transcode(nBlocks) === 0 ? textureView.slice() : null;
+        }, (reason) => {
+            throw new Error(reason);
+        });

```

Comment:
The only case when `transcode()`'s return value is not zero is when the `nBlocks` value is negative or out of bounds to avoid wasm runtime exception. If the code calling `transcode()` guaranties that `nBlocks` value fits the allocated memory, the error check could be omitted.

In a case when the texture data is invalid, the transcoding succeeds with invalid blocks being replaced with `[255, 0, 255, 255]` color.

Reply:
Even if not strictly required by the comment from @lexaknyazev, I feel being a bit defensive should not harm?

We never know how the code could evolve in the future.

Reply:
I should probably be more clear. I'm asking why we need to catch and rethrow the error when this function is returning a promise. That is not how promise returning functions work normally.

Reply:
I have removed this handler and the one in the msc transcoder.